diesel/src/connection/transaction_manager.rs

```diff
@@ -237,55 +285,92 @@ where
-                            rollback_result,
-                        );
-                        Err(Error::CommitTransactionFailed {
-                            commit_error: Box::new(error),
-                            rollback_result: Box::new(rollback_result),
-                        })
-                    }
-                    CommitErrorOutcome::Throw(error) => {
-                        // The error processor indicated that we just
-                        // need to decrease the transaction depth and return the original error
-                        let _ = state.change_transaction_depth(
-                            TransactionDepthChange::DecreaseDepth,
-                            Ok(()),
-                        );
-                        Err(Error::CommitTransactionFailed {
-                            commit_error: Box::new(error),
-                            rollback_result: Box::new(Ok(())),
-                        })
-                    }
-                    CommitErrorOutcome::ThrowAndMarkManagerAsBroken(error) => {
-                        // The connection contains an unrecoverable broken transaction
-                        // so mark the transaction state as broken and return the error
-                        *Self::transaction_manager_status_mut(conn) =
-                            TransactionManagerStatus::InError;
-                        Err(Error::CommitTransactionFailed {
-                            commit_error: Box::new(error),
-                            rollback_result: Box::new(Err(Error::BrokenTransaction)),
-                        })
-                    }
-                }
-            }
+            Err(error) => process_commit_error(conn, error, rollback_sql),
         }
     }
 
     fn transaction_manager_status_mut(conn: &mut Conn) -> &mut TransactionManagerStatus {
         &mut conn.transaction_state().status
     }
+
+    fn transaction<F, R, E>(conn: &mut Conn, f: F) -> Result<R, E>
+    where
+        F: FnOnce(&mut Conn) -> Result<R, E>,
+        E: From<Error>,
+    {
+        Self::begin_transaction(conn)?;
+        match f(&mut *conn) {
+            Ok(value) => {
+                Self::commit_transaction(conn)?;
+                Ok(value)
+            }
+            Err(e) => {
+                let transaction_state = Self::get_transaction_state(conn)?;
+                let is_serialization_error = transaction_state
+                    .previous_error_relevant_for_rollback
+                    .is_some();
+
+                Self::rollback_transaction(conn).map_err(|e| {
+                    if is_serialization_error {
+                        if let Error::DatabaseError(
+                            crate::result::DatabaseErrorKind::SerializationFailure,
+                            _,
+                        ) = e
+                        {
+                            return e;
+                        }
+                    }
```

Comment:
More for my own understanding that anything, since we decided to introduce a generic error relevant for rollback I wonder two things
1. Should `is_serialization_error` be named something else like `is_error_relevant_for_rollback`?
2. Should the matching on line 315-321 just forward the error kind of the `Some(transaction_state. previous_error_relevant_for_rollback)` (I think it's a tuple so maybe can just return the `.0`th element of that `Some()`)

Asking because it seems to me like we should just forward along anything inside the `if` check on line 314 regardless of whether it matches a `crate::result::DatabaseErrorKind::SerializationFailure` or not

Reply:
> 1. Should is_serialization_error be named something else like is_error_relevant_for_rollback?

Yes that should be named different. That's a left over part of the old verison.

> 2. Should the matching on line 315-321 just forward the error kind of the Some(transaction_state. previous_error_relevant_for_rollback) (I think it's a tuple so maybe can just return the .0th element of that Some())

This error mapping here is likely not even required as `rollback_transaction` will handle that internally.

Reply:
Ah I see! In that case it seems like this could all be simplified by just calling:
```
Err(e) => {
    let transaction_state = Self::get_transaction_state(conn)?;
    Self::rollback_transaction(conn).map_err(|e| Error::RollbackError(Box::new(e)))?;
    Err(e)
}
```

Anyway thanks for fielding my questions!

Reply:
So I had a look at the corresponding code again. Seems like we cannot simplify that anymore, beside the name change. ` Self::rollback_transaction(conn).map_err(|e| Error::RollbackError(Box::new(e)))?;` does not work as we want to return a `Error::DatabaseError(crate::result::DatabaseErrorKind::SerializationFailure,  _)` for special cases, not a `RollbackError`. Both, the `is_serialization_error` and the actual error checking are there as a safeguard, so that we do only not wrap the inner error in those cases.

I've filed https://github.com/diesel-rs/diesel/pull/3128 for the variable name.

Reply:
Ah! You know what it was.. I was thrown off by the `SerializationFailure` as the hardcoded error type (albeit within a guard clause) because I recall originally you had asked whether this error type should be made more generic.

So I guess my question now is why not just return the error `e` without the guard clause since we already know it's relevant for the rollback (generic `DatabaseErrorKind ` error)?
```rust
Err(e) => {
    let transaction_state = Self::get_transaction_state(conn)?;
    let is_error_relevant_for_rollback = transaction_state
        .previous_error_relevant_for_rollback
        .is_some();

    Self::rollback_transaction(conn).map_err(|e| {
        if is_error_relevant_for_rollback {
            return e;
        }

        Error::RollbackError(Box::new(e))
    })?;
    Err(e)
}
```